engine/server/engine/server/enclave_rest_api_handler.go

```diff
@@ -0,0 +1,857 @@
+// ===============================================================================================================
+
+// GetGrpcClientConn returns a client conn dialed in to the local port
+// It is the caller's responsibility to call resultClientConn.close()
+func getGrpcClientConn(enclaveInfo types.EnclaveInfo, connectOnHostMachine bool) (resultClientConn *grpc.ClientConn, resultErr error) {
+	enclaveAPIContainerInfo := enclaveInfo.ApiContainerInfo
+	if enclaveAPIContainerInfo == nil {
+		logrus.Infof("No API container info is available for enclave %s", enclaveInfo.EnclaveUuid)
+		return nil, nil
+	}
+	apiContainerGrpcPort := enclaveAPIContainerInfo.GrpcPortInsideEnclave
+	apiContainerIP := enclaveAPIContainerInfo.BridgeIpAddress
+
+	enclaveAPIContainerHostMachineInfo := enclaveInfo.ApiContainerHostMachineInfo
+	if connectOnHostMachine && enclaveAPIContainerHostMachineInfo == nil {
+		logrus.Infof("No API container info is available for enclave %s", enclaveInfo.EnclaveUuid)
+		return nil, nil
+	}
+	if connectOnHostMachine && (enclaveAPIContainerHostMachineInfo != nil) {
+		apiContainerGrpcPort = enclaveAPIContainerHostMachineInfo.GrpcPortOnHostMachine
+		apiContainerIP = enclaveAPIContainerHostMachineInfo.IpOnHostMachine
+	}
+
+	grpcServerAddress := fmt.Sprintf("%v:%v", apiContainerIP, apiContainerGrpcPort)
+	grpcConnection, err := grpc.Dial(grpcServerAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "Expected to be able to create a GRPC client connection on address '%v', but a non-nil error was returned", grpcServerAddress)
+	}
+	return grpcConnection, nil
+}
+
+func (manager enclaveRuntime) GetApiClientOrResponseError(enclave_uuid string) (*rpc_api.ApiContainerServiceClient, *api_type.ResponseInfo) {
+	client, err := manager.GetGrpcClientForEnclaveUUID(enclave_uuid)
+	if err != nil {
+		return nil, &api_type.ResponseInfo{
+			Type:    api_type.ERROR,
+			Message: "Couldn't retrieve connection with enclave",
+			Code:    http.StatusInternalServerError,
+		}
+	}
+	if client == nil {
+		return nil, &api_type.ResponseInfo{
+			Type:    api_type.INFO,
+			Message: fmt.Sprintf("enclave '%s' not found", enclave_uuid),
+			Code:    http.StatusNotFound,
+		}
+	}
+	return client, nil
+}
+
+func (manager enclaveRuntime) GetGrpcClientForEnclaveUUID(enclave_uuid string) (*rpc_api.ApiContainerServiceClient, error) {
+	err := manager.refreshEnclaveConnections()
+	if err != nil {
+		return nil, err
+	}
+
+	client, found := manager.remoteApiContainerClient[enclave_uuid]
+	if !found {
+		return nil, nil
+	}
+
+	return &client, nil
+}
+
+func toGrpcConnect(conn api_type.Connect) rpc_api.Connect {
+	switch conn {
+	case api_type.CONNECT:
+		return rpc_api.Connect_CONNECT
+	case api_type.NOCONNECT:
+		return rpc_api.Connect_NO_CONNECT
+	default:
+		panic(fmt.Sprintf("Missing conversion of Connect Enum value: %s", conn))
```

Comment:
I suggest returning an error instead of panicking here

Reply:
this is one of the panicking could make sense but we don't expect a failure under normal operations. This is an on-to-one mapping between enums and the only failure mode is if one adds a new value to the enum. I'm fine adding code to handle it throughout the codebase (as we need to keep passing it through the function stack calls) but it would be great if Go would have a native concept of enum exhaustion in switch statements (maybe we can work to add smtg like https://pkg.go.dev/github.com/nishanths/exhaustive to enforce that check).

Reply:
I've added the [exhaustive](https://pkg.go.dev/github.com/nishanths/exhaustive) to our golangci-lint and the default case should never be reached.